CurrencyConverterControl/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,334 @@
+                {
+                    return;
+                }
+                control.OutputValue = CurrencyDataProvider.ConvertAsync(control.SourceCurrency,
+                                                            control.DestinationCurrency, control.InputValue).Result;               
+                
+            }
+            catch (Exception ex) { Console.WriteLine($"Error: {ex.Message}"); }
+        }
+
+        /// <summary>
+        /// Gets the Source Currency to perform the currency conversion  
+        /// </summary>
+        [Description("Value of the source currency")]
+        public string SourceCurrency
+        {
+            get { return (string)GetValue(SourceCurrencyProperty); }
+            set 
+            { 
+                SetValue(SourceCurrencyProperty, value);
+                
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the ObservableCollection of Currency objects.
+        /// </summary>
+        /// <remarks>
+        /// The CurrencyList property represents a collection of Currency objects,
+        /// typically used for storing a list of currencies available in the application
+        /// to fill in the ComboBox element.
+        /// It is initialized with a new instance of ObservableCollection<Currency>.
+        /// </remarks>
+        [Description("Collection of currencies list")]
+        public ObservableCollection<Currency> CurrencyList { get; private set; } = new ObservableCollection<Currency>();
+
+        /// <summary>
+        /// Property to allow replacement of the currency converter
+        /// </summary>
+        public ICurrencyDataProvider CurrencyConverter
+        {
+            get { return CurrencyDataProvider; }
+            set { CurrencyDataProvider = value ?? throw new ArgumentNullException(nameof(value)); }
+        }
+
+        public static readonly DependencyProperty CustomStyleProperty =
+            DependencyProperty.Register(nameof(CustomStyle), typeof(Style), typeof(CurrencyConverterControl));
+        /// <summary>
+        /// Define custom styles on the control
+        /// </summary>
+        public Style CustomStyle
+        {
+            get { return (Style)GetValue(CustomStyleProperty); }
+            set { SetValue(CustomStyleProperty, value); }
+        }
+        #endregion
+        public override void OnApplyTemplate()
+        {
+
+            LoadCurrencies();
+            CurrencyDataProvider.GetConvertionRates();
+            _currencyDestination = Template.FindName("CmbDestination", this) as ComboBox;
+            _currencySource = Template.FindName("CmbSource", this) as ComboBox;
+            //To Be done
+            if (_currencyDestination is null || _currencySource is null)
+            { throw new InvalidOperationException("The Combobox in the Template is not found! Please double check!"); }
+            else 
+            {
```

Comment:
OnApplyTemplate might be called many times during application life time. So, the first thing it should do - is check whether _currencyDestination and other parts are not null and unsubscribe from events. Then search for new control and subscribe. If you don't unsubscribe, dead unloaded controls will still have links to CurrencyConverterControl and this will cause memory leaks